src/jit/morph.cpp

```diff
@@ -9315,232 +9314,260 @@ GenTree* Compiler::fgMorphInitBlock(GenTree* tree)
 #endif // LOCAL_ASSERTION_PROP
 
-            if (destLclVar->lvPromoted && blockWidthIsConst)
+            if (destLclVar->lvPromoted)
             {
-                assert(initVal->OperGet() == GT_CNS_INT);
-                noway_assert(varTypeIsStruct(destLclVar));
-                noway_assert(!opts.MinOpts());
-                if (destLclVar->lvAddrExposed & destLclVar->lvContainsHoles)
-                {
-                    JITDUMP(" dest is address exposed");
-                }
-                else
+                GenTree* newTree = fgMorphPromoteLocalInitBlock(destLclNode->AsLclVar(), initVal, blockSize);
+
+                if (newTree != nullptr)
                 {
-                    if (blockWidth == destLclVar->lvExactSize)
-                    {
-                        JITDUMP(" (destDoFldAsg=true)");
-                        // We may decide later that a copyblk is required when this struct has holes
-                        destDoFldAsg = true;
-                    }
-                    else
-                    {
-                        JITDUMP(" with mismatched size");
-                    }
+                    tree         = newTree;
+                    destDoFldAsg = true;
+                    INDEBUG(morphed = true);
                 }
             }
-        }
-
-        // Can we use field by field assignment for the dest?
-        if (destDoFldAsg && destLclVar->lvCustomLayout && destLclVar->lvContainsHoles)
-        {
-            JITDUMP(" dest contains holes");
-            destDoFldAsg = false;
-        }
 
-        JITDUMP(destDoFldAsg ? " using field by field initialization.\n" : " this requires an InitBlock.\n");
-
-        // If we're doing an InitBlock and we've transformed the dest to a non-Blk
-        // we need to change it back.
-        if (!destDoFldAsg && !dest->OperIsBlk())
-        {
-            noway_assert(blockWidth != 0);
-            tree->gtOp.gtOp1 = origDest;
-            tree->gtType     = origDest->gtType;
-        }
-
-        if (!destDoFldAsg && (destLclVar != nullptr))
-        {
             // If destLclVar is not a reg-sized non-field-addressed struct, set it as DoNotEnregister.
-            if (!destLclVar->lvRegStruct)
+            if (!destDoFldAsg && !destLclVar->lvRegStruct)
             {
-                // Mark it as DoNotEnregister.
                 lvaSetVarDoNotEnregister(destLclNum DEBUGARG(DNER_BlockOp));
             }
         }
 
-        // Mark the dest struct as DoNotEnreg
-        // when they are LclVar structs and we are using a CopyBlock
-        // or the struct is not promoted
-        //
         if (!destDoFldAsg)
         {
-            dest             = fgMorphBlockOperand(dest, dest->TypeGet(), blockWidth, true);
+            // If we're doing an InitBlock and we've transformed the dest to a non-Blk
+            // we need to change it back.
+            if (!dest->OperIsBlk())
+            {
+                noway_assert(blockSize != 0);
```

Comment:
when does it happen? When we call `fgMorphPromoteLocalInitBlock` it changes `dest` but returns `null`?

Reply:
Nah, `fgMorphPromoteLocalInitBlock` doesn't touch the tree if it returns null. This is due to `fgMorphBlkNode` that's called at the start of `fgMorphInitBlock`. Apparently `fgMorphBlkNode` can sometimes transform the block node to a `GT_IND` that's obviously not suitable as a struct store destination because it doesn't have a size. So if we fail to perform field by field initialization we need to change it back.

Now, don't ask me why `fgMorphBlkNode` produces such a `GT_IND` node... Struct promotion and more generally struct handling in the JIT needs a serious shake up. I'll get to it eventually, need to finish #21705 first.

Reply:
> Maybe we can change fgMorphBlkNode not to do this?

I hope so. But this needs to be investigated separately because it's pretty convoluted.

As far as I can tell, `fgMorphBlkNode` can only return `GT_IND` if it can't find a class handle. And it only looks for a class handle if it deals with an array element access and then a class handle should exist on `GT_INDEX`/`GT_INDEX_ADDR`.

Except the odd corner case when the destination happens to be a `GT_COMMA` that's not actually related to array access. If that ever happens then that assert in `fgMorphInitBlock` will fire. It really shouldn't happen because the importer is pretty insistent in wrapping struct trees with `GT_OBJ`/`GT_BLK` nodes and there's a good chance that the `GT_COMMA` (and likely the `GT_INDEX_ADDR`) `fgMorphBlkNode` code is actually dead.

Reply:
Thanks for the explanation!